udf/mrn_udf_highlight_html.cpp

```diff
@@ -356,26 +413,41 @@ static bool highlight_html(grn_ctx *ctx,
                                 hits, MAX_N_HITS, &rest);
       for (int i = 0; i < n_hits; i++) {
         if ((hits[i].offset - previous) > 0) {
-          grn_text_escape_xml(ctx,
-                              output,
-                              target + previous,
-                              hits[i].offset - previous);
+          highlight_html_put_text(ctx,
+                                  output,
+                                  target + previous,
+                                  hits[i].offset - previous,
+                                  need_escape);
+        }
+        if (skip_tagging_in_char_ref) {
+          in_char_ref = 
+            highlight_html_check_in_char_ref(ctx,
```

Comment:
How about checking `&...;` in `mroonga_highlight_html()` too like `<...>`?

Reply:
If we check `&..;` in `mroonga_highlight_html()` and simply skip it, we can not highlight a keyword that contains `&..;` like the following test.
I think we should check whether the current position is in `&..;` after detecting keywords from a target text.

https://github.com/mroonga/mroonga/pull/540/files#diff-9fea1f52acf8445e116584bf15b459fae19c19eeb23a6973e262b3184927f17aR1

Reply:
I think that we can to check whether in the current position is  `<...>` in `highlight_html`, but it may complicate the conditional branching of `highlight_html`. So I decided to check `<...>` in `mroonga_highlight_html()`.

Reply:
Hmm, what is the expected result for `aaa<b>hello</b>bbb` as the input and `<b>hello</b>` as a keyword?
`aaa<span class="keyword"><b>hello</b></span>bbb`? `aaa<b>hello</b>bbb` (same as the input text) ?

Reply:
In my opinion, users should specify `text` as the input type if a keyword contains tags (`<>`) because `<>` is not appeared inside '<>' and mishighlighting inside `<>` doesn't occur.
But I'm agree with that this is little confusing... Hmm...

Reply:
Ah, I'm wrong, if the input type is `text`, `<` and so on escaped...

Reply:
> Hmm, what is the expected result for `aaa<b>hello</b>bbb` as the input and `<b>hello</b>` as a keyword?
`aaa<span class="keyword"><b>hello</b></span>bbb`? `aaa<b>hello</b>bbb` (same as the input text) ?

I have fixed to get `aaa<span class="keyword"><b>hello</b></span>bbb` when a user specify `aaa<b>hello</b>bbb` as the input and `<b>hello</b>` as a keyword.

Reply:
Ah, if an user specifies `<b>hello` as a keyword, it will generate a broken HTML...

Reply:
Yes, but I think it is an user's responsibility to specify a correct keyword.
How do you think to just adding a notification about this for the `mroonga_highlight_html` document for now and fixing it when we need to fix it in the future?

Reply:
Could you summarize the current specification? I want to confirm whether there is inconsistency or not.

Reply:
> Could you summarize the current specification? I want to confirm whether there is inconsistency or not.

I have added the current specification into the description of this PR.
https://github.com/mroonga/mroonga/pull/540#issue-1374262373

Please confirm it.

Reply:
Considering use cases for this feature, the following specification seems hard to use:

> * We can use character references or HTML tags for keywords
>   * if we specify `<span class="abc">abc<br/>def</span>` as target text and `c<br/>d` as keyword, the result is `<span class="abc">ab<span class="keyword">c<br/>d</span>ef</span>`.
>   * if we specify `<span class="abc">abc&amp;def</span>` as target text and `c&amp;d` as keyword, the result is `<span class="abc">ab<span class="keyword">c&amp;d</span>ef</span>`.
>
> Note that if an user specifies a target text like `<b>abc</b>` and a keyword like `abc</b>`, the result is `<b><span class="keyword">abc</d></span>` and it is broken html.
> But it is user's responsibility to specify correct keywords.

An use case (see #536):
* Extract HTML tags from target texts manually (in an app) and put the extracted text to Groonga
* Search an input from end-users (not app developers) against the extracted text
* Use the input from end-users (not app developers) to highlight the original HTML included text (not the extracted text)

In this use case, the target extracted text doesn't have HTML tags. So `c<br/>d` isn't matched. So it should not be highlighted.

In this use case, keywords for this context are specified by end-users not app developers. So "it is user's responsibility to specify correct keywords" is hard. Because app developers can't control it.

How about the followings?

* We can use HTML tags for keywords but they are never matched
* We can use character references (bonus)
  * We expand character references in target text and keyword before we process: `&amp;` -> `&`
  * We highlight with the expanded target text and the expanded keyword, and then escape the result like `AS text` case:
    * target text: `a &amp; b &#38; c`
      * keyword: `&` -> `a <span class="keyword">&amp;</span> b <span class="keyword">&amp;<span> c`
      * keyword: `&amp;` -> `a <span class="keyword">&amp;</span> b <span class="keyword">&amp;<span> c`
* Highlighted HTML isn't broken if the target text isn't broken

Reply:
Thanks.

>  * We expand character references in target text and keyword before we process: `&amp;` -> `&`

I think we need to support all of character references for it.

https://html.spec.whatwg.org/multipage/named-characters.html

Are there good libraries or functions for it?

Reply:
We can use them statically because it'll not be changed:

> This list is static and [will not be expanded or changed in the future](https://github.com/whatwg/html/blob/main/FAQ.md#html-should-add-more-named-character-references).

I think that using an external library is overkill for our use case.